lib/tests/test_rewrite.rs

```diff
@@ -1889,3 +1893,146 @@ fn test_commit_with_selection() {
     assert!(!full_selection.is_empty_selection());
     assert!(full_selection.is_full_selection());
 }
+
+#[test]
+fn test_find_duplicate_divergent_commits() {
+    let test_repo = TestRepo::init();
+    let repo = &test_repo.repo;
+
+    let mut tx = repo.start_transaction();
+    let store = repo.store();
+    // We want to create the following tree with divergent changes A, B, and C:
+    //     E
+    //     |
+    // A2  C2
+    // |   |
+    // C1  B2
+    // |   |
+    // B1  D
+    // |  /
+    // | /
+    // A1
+
+    let tree_a1 = create_tree(repo, &[(repo_path("file1"), "a\n")]);
+    // Commit b deletes file1 and adds file2
+    let tree_b1 = create_tree(repo, &[(repo_path("file2"), "b\n")]);
+    // Commit c appends to file2
+    let tree_c1 = create_tree(repo, &[(repo_path("file2"), "b\nc\n")]);
+    // Commit a2 re-adds file1, the same as commit a1. Since it already has a1
+    // as an ancestor, it shouldn't be abandoned even though it is divergent.
+    let tree_a2 = create_tree(
+        repo,
+        &[(repo_path("file1"), "a\n"), (repo_path("file2"), "b\nc\n")],
+    );
+    let tree_d = create_tree(
+        repo,
+        &[(repo_path("file1"), "a\n"), (repo_path("file3"), "d\n")],
+    );
+    let tree_b2 = create_tree(
+        repo,
+        &[(repo_path("file2"), "b\n"), (repo_path("file3"), "d\n")],
+    );
+    let tree_c2 = create_tree(
+        repo,
+        &[(repo_path("file2"), "b\nc\n"), (repo_path("file3"), "d\n")],
+    );
+    let tree_e = create_tree(
+        repo,
+        &[
+            (repo_path("file2"), "b\nc\n"),
+            (repo_path("file3"), "d\ne\n"),
+        ],
+    );
+
+    let mut make_commit = |tree_id, change_id, parents| {
+        tx.repo_mut()
+            .new_commit(parents, tree_id)
+            .set_change_id(change_id)
+            .write()
+            .unwrap()
+    };
+
+    let commit_a1 = make_commit(
+        tree_a1.id(),
+        ChangeId::from_hex("aaaa"),
```

Comment:
nit: probably better to generate a `ChangeId` of valid length because it is stored in the test backend.